Differentiate prelude constructor and custom type constructors

shunjilin opened the pull request:
closes #1112 

Add `type_` field to `Pattern` constructor to allow js to differentiate between prelude constructor and custom type constructor

shunjilin commented on src/parse.rs:
```diff
@@ -1040,6 +1040,7 @@ where
             name,
             with_spread,
             constructor: (),
+            type_: (),
```
Looking at the signature I assumed this is untyped, not sure if putting `()` here is correct

lpil reviewed (commented):
Looks great!

As a bit of a nit-pick can we keep difference match clauses for true, false, and nil that isn't with the other record clause? I find it clearer if it's split out.

We could use something like this:

```rust
            Pattern::Constructor {
                constructor: PatternConstructor::Record { type_, .. },
                ..
            } if type_.is_nil() => {
                self.booly_check(false);
                Ok(())
            }
```

shunjilin commented:
> 
> 
> Looks great!
> 
> As a bit of a nit-pick can we keep difference match clauses for true, false, and nil that isn't with the other record clause? I find it clearer if it's split out.
> 
> We could use something like this:
> 
> ```rust
>             Pattern::Constructor {
>                 constructor: PatternConstructor::Record { type_, .. },
>                 ..
>             } if type_.is_nil() => {
>                 self.booly_check(false);
>                 Ok(())
>             }
> ```

Thank you! Fixed, and looks cleaner indeed

> 
> 
> Looks great!
> 
> As a bit of a nit-pick can we keep difference match clauses for true, false, and nil that isn't with the other record clause? I find it clearer if it's split out.
> 
> We could use something like this:
> 
> ```rust
>             Pattern::Constructor {
>                 constructor: PatternConstructor::Record { type_, .. },
>                 ..
>             } if type_.is_nil() => {
>                 self.booly_check(false);
>                 Ok(())
>             }
> ```

Thank you! Fixed, and looks cleaner indeed

lpil reviewed (approved):
Wonderful work! Thank you! ✨

lpil commented:
CI is broken because of a mistake I made, can ignore that
